File: src/RealWorld.Application/Pagination/CursorPager.cs
namespace RealWorld.Application.Pagination;

public class CursorPager<T> where T : INode
{
    public List<T> Data { get; }
    public bool Next { get; }
    public bool Previous { get; }

    public CursorPager(List<T> data, Direction direction, bool hasExtra)
    {
        Data = data;
        if (direction == Direction.Next)
        {
            Previous = false;
            Next = hasExtra;
        }
        else
        {
            Next = false;
            Previous = hasExtra;
        }
    }

    public bool HasNext() => Next;
    public bool HasPrevious() => Previous;

    public PageCursor<DateTimeOffset>? GetStartCursor()
    {
        return Data.Count == 0 ? null : Data[0].GetCursor();
    }

    public PageCursor<DateTimeOffset>? GetEndCursor()
    {
        return Data.Count == 0 ? null : Data[^1].GetCursor();
    }
}

public enum Direction
{
    Prev,
    Next
}

File: article-service/src/main/java/io/spring/article/application/CursorPager.java
package io.spring.article.application;

import java.util.List;
import lombok.Getter;

@Getter
public class CursorPager<T extends Node> {
  private List<T> data;
  private boolean hasExtra;
  private CursorPageParameter.Direction direction;

  public CursorPager(List<T> data, CursorPageParameter.Direction direction, boolean hasExtra) {
    this.data = data;
    this.direction = direction;
    this.hasExtra = hasExtra;
  }

  public PageCursor<T> getStartCursor() {
    return data.isEmpty() ? null : new PageCursor<>(data.get(0));
  }

  public PageCursor<T> getEndCursor() {
    return data.isEmpty() ? null : new PageCursor<>(data.get(data.size() - 1));
  }

  public boolean hasNextPage() {
    if (direction == CursorPageParameter.Direction.NEXT) {
      return hasExtra;
    }
    return true;
  }

  public boolean hasPreviousPage() {
    if (direction == CursorPageParameter.Direction.PREV) {
      return hasExtra;
    }
    return true;
  }
}
